src/Compilers/Core/Portable/Compilation.EmitStream.cs

```diff
@@ -0,0 +1,176 @@
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    public abstract partial class Compilation
+    {
+        /// <summary>
+        /// Describes the kind of real signing that is being done during Emit. In the case of public signing
+        /// this value will be <see cref="None"/>.
+        /// </summary>
+        internal enum EmitStreamSignKind
+        {
+            None,
+            SignedWithBulider,
+            SignedWithFile,
+        }
+
+        /// <summary>
+        /// This type abstracts away the legacy COM based signing implementation for PE streams. Under the hood
+        /// a temporary file must be created on disk (at the last possible moment), emitted to, signed on disk
+        /// and then copied back to the original <see cref="Stream"/>. Only when legacy signing is enabled though.
+        /// </summary>
+        internal sealed class EmitStream
+        {
+            private readonly EmitStreamProvider _emitStreamProvider;
+            private readonly EmitStreamSignKind _emitStreamSignKind;
+            private readonly StrongNameProvider _strongNameProvider;
+            private (Stream tempStream, string tempFilePath)? _tempInfo;
+
+            /// <summary>
+            /// The <see cref="Stream"/> that is being emitted into. This value should _never_ be 
+            /// Dispose. It is either returned from the <see cref="EmitStreamProvider"/> instance in
+            /// which case it is owned by that. Or it is just an alias for the value that is stored 
+            /// in <see cref="_tempInfo"/> in which case it will be disposed from there.
+            /// </summary>
+            private Stream _stream;
+
+            internal EmitStream(
+                EmitStreamProvider emitStreamProvider,
+                EmitStreamSignKind emitStreamSignKind,
+                StrongNameProvider strongNameProvider)
+            {
+                Debug.Assert(emitStreamProvider != null);
+                Debug.Assert(strongNameProvider != null || emitStreamSignKind == EmitStreamSignKind.None);
+                _emitStreamProvider = emitStreamProvider;
+                _emitStreamSignKind = emitStreamSignKind;
+                _strongNameProvider = strongNameProvider;
+            }
+
+            internal Func<Stream> GetCreateStreamFunc(DiagnosticBag diagnostics)
+            {
+                return () => CreateStream(diagnostics);
+            }
+
+            internal void Dispose()
+            {
+                _tempInfo?.tempStream.Dispose();
+                _tempInfo = null;
+
+                // The _stream value is deliberately excluded from being disposed here. That value is not 
+                // owned by this type.
+                _stream = null;
+            }
+
+            /// <summary>
+            /// Create the stream which should be used for Emit. This should only be called one time.
```

Comment:
>This should only be called one time. [](start = 65, length = 36)

Consider adding some debug asserts that this is the case.

Reply:
The assert for `stream == null` is meant to check this. It does have the problem that if you call `CreateStream`, then `Dispose` then `CreateStream` that the assert will fail though. 

---
In reply to: [245373577](https://github.com/dotnet/roslyn/pull/31986#discussion_r245373577) [](ancestors = 245373577)